plugins/inputs/p4runtime/README.md

```diff
@@ -0,0 +1,79 @@
+# P4 Runtime Input Plugin
+
+The `p4runtime` plugin gathers metrics about `Counter` values
+present in P4 Program loaded onto Programmable Switch.
+Metrics are collected through gRPC connection with
+[P4Runtime](https://github.com/p4lang/p4runtime) server.
+
+P4Runtime Plugin uses `PkgInfo.Name` field.
+If user want to gather information about program name, please follow
+[6.2.1. Annotating P4 code with PkgInfo] in P4 program.
+
+[6.2.1. Annotating P4 code with PkgInfo]: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-annotating-p4-code-with-pkginfo
+
+## Configuration
+
+```toml @sample.conf
+# P4Runtime telemetry input plugin
+[[inputs.p4runtime]]
+  ## Define the address of P4Runtime gRPC server to collect metrics.
+  # endpoint = "127.0.0.1:9559"
+  ## Set DeviceID required for Client Arbitration.
+  ## https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-client-arbitration-and-controller-replication 
+  # device_id = "1"
+  ## Filter counters by their names that should be observed.
+  ## Example: counter_names=["ingressCounter", "egressCounter"]
+  # counter_names = []
+
+  ## Optional TLS Config.
+  ## Enable client-side TLS and define CA to authenticate the device.
+  # enable_tls = false
+  # tls_ca = "/etc/telegraf/ca.crt"
+  ## Set minimal TLS version to accept by the client.
+  # tls_min_version = "TLS12"
+  ## Use TLS but skip chain & host verification.
+  # insecure_skip_verify = true
+
+  ## Define client-side TLS certificate & key to authenticate to the device.
+  # tls_cert = "/etc/telegraf/client.crt"
+  # tls_key = "/etc/telegraf/client.key"
+```
+
+## Metrics
+
+P4Runtime gRPC server communicates using [p4runtime.proto](
+https://github.com/p4lang/p4runtime/blob/main/proto/p4/v1/p4runtime.proto)
+Protocol Buffer.
+Static information about P4 program loaded into programmable switch
+are collected by `GetForwardingPipelineConfigRequest` message.
+Plugin gathers dynamic metrics with `Read` method.
+`Readrequest` is defined with single `Entity` of type `CounterEntry`.
+Since P4 Counter is array, plugin collects values of every cell of array
+by [wildcard query].
+
+[wildcard query]: https://github.com/p4lang/p4runtime/blob/main/proto/p4/v1/p4runtime.proto#L379
+
+Counters defined in P4 Program have unique ID and name.
+Counters are arrays, thus `counter_index` informs
+which cell value of array is described in metric.
+
+Tags are constructed in given manner:
+
+`p4program_name`: P4 program name provided by user.
+Instruction [6.2.1. Annotating P4 code with PkgInfo]
+`counter_name`: Name of given counter in P4 program.
+`counter_index`: Index at which metrics are collected in P4 counter.
```

Comment:
Having an index be a tag is usually not ideal. However in this use case, is this a bound number? or can it grow forever? 

Is it helpful to an end user to have this as tag to filter on?

Reply:
no code change, just answer the comment

Indirect counters are arrays, with max size of 32 bits. Counting is done with a .count(index) method - so the increment happens on a specific index of the array. That means one counter can have several count methods done on it, each on a different index.
For some insight, index could be for example a port number.
It is bound to 32 bits and it won't grow forever.

Reply:
> It is bound to 32 bits

That is still a pretty large number :) @zak-pawel should this really be a tag? Should this even be collected at all?

Reply:
I understand your confusion, please let's have a look into an example why we need this tag (based on [p4 lang exercise](https://github.com/p4lang/tutorials/tree/master/exercises/p4runtime), similar example also available in [ipdk](https://github.com/ipdk-io/networking-recipe)).



In this example we will have 3 virtual machines - named `h1`, `h2` and `h3`. Each of these VMs will have it's own P4 programmable switch - `s1`, `s2` and `s3`.

We will install a P4 program on switches `s1` and `s2` and write rules that tunnel traffic between their corresponding VMs. `S3` will not have a P4 program or any traffic rules.


**Example 1.**

Sending traffic from `h1` to `h2`:



This is what is visible from the plugin after connecting to switch 1:

![MicrosoftTeams-image (11)](https://user-images.githubusercontent.com/74768577/215453147-264ef361-f7fb-486c-ae46-8ce48b62c557.png)



2 Counters Ingress ("from where") and Egress ("where to") with 2 different `counter_index` values.

First line - ingress counter with index 100 - tells us how much traffic was sent from tunnel with ID 100 - which in this case is VM named `h1`.

Second line - egress counter with index 200 - tells us how much traffic was sent to tunnel with ID 200 - which in this case is VM named `h2`.

These IDs come from the rules we created for each switch.



From just these 2 lines we can tell from where we send traffic (VM named `h1`), where it's sent to (VM named `h2`) and because both of these counters packets are equal - there was no packet loss. We know that because we have `counter_index` values (and understand traffic rules).




**Example 2.**

Sending traffic from `h1` to `h3`.

This is what is visible from plugin after connecting to switch 1:

![MicrosoftTeams-image (12)](https://user-images.githubusercontent.com/74768577/215453201-f1790979-bd6d-4e16-be2c-776b0156870a.png)



As you can see we only get ingress counter here - we're sending traffic from tunnel with ID 100 (so from VM `h1`). No egress counter here (with id 300) however, because it's values are 0 - we're not receiving any traffic on VM `h3` (because we didn't create rules for that).



We can see there is data being sent - but there is no receiving end - p4 flow works, but it lacks a rule.



I hope based on this simple example you can see why we need the `counter_index` tag now.





As for the size of this tag - it depends on your p4 programmable pipeline - how many rules you create and more importantly, how many of them you want to monitor this way. In this example, these switches are used by VMs - so the amount is bound by how many VMs you can create. This could be containers. Or it could be bound by hardware. None of these will lead to it becoming an excessively large number.

Reply:
Thanks for this, this is helpful to see a real example.

> These IDs come from the rules we created for each switch.

Is there any other, more identifiable tag? 

For example, does the P4 runtime have anyway of knowing the VM hostname and instead of an ID return the hostname? Would this also make it easier on users when trying to analyze the results rather than needing to go to the switch to translate what index belongs to what host?

Reply:
There isn't any other, more identifiable tag that we could extract from P4 runtime. There's nothing more in P4Info (from P4 Runtime) that could help us here.

P4 is a network programming language. So it's all about packet processing on different data plane devices - routers, sw & hw switches, NICs. So it doesn't really understand which devices it is working on - all it cares about are packets and the rules we create to handle them on these devices (which we identify by IDs or "ports" - `counter_index`).

Reply:
My concern is that the carnality of an InfluxDB bucket would be huge with this plugin as-is. An index that can grow 2 billion index * counter name * counter type * hosts is too large of a potential cardinality.

Can we move the index to a field please? Then this makes using the index as a tag as an opt-in process via the converter processor.